[WIP] Red's launcher / Downloader reqs

Twentysix26 opened the pull request:
**Launcher:**
- A launcher that deals with requirements, updates and maintenance
- Interactive with menus / submenus, etc but can also be headless with CLI flags
- Brings the startRed.bat / startRedLoop.bat functionality on all platforms
- Deals with requirements, installs them locally in the local `lib` folder without the need of root
- Plenty of updates options
- Downloads ffmpeg on Windows and verifies its files' integrity
- Maintenance options: can reset Red, its requiremenents and factory reset
- Generates batch files for fast start on Windows

**Downloader changes:**
- Cog-specific `info.json` can now have a `"REQUIREMENTS"` field in which they specify their requirements
- The requirements will be installed via pip in a thread safe way, just before the cog gets installed. The info.json will be verified after each update for any new reqs
- Since the requirements will be installed inside the local `lib` folder it won't require root
- Failure to install the requirements will make the cog installation fail

**Other changes:**
- A `[p]restart` command has been added. It makes Red quit with exit code 26. Once receiving that code the launcher will start it up again
- In the launcher context, `[p]shutdown` will actually quit Red for good, regardless of the autorestart mode
- `[p]shutdown` and `[p]restart` now both use the new `Bot.shutdown` method

**Breaking changes:**
- In theory, nothing. 🤞

**Issues:**
- ~~Currently not working on Macs. Possibly related to [this](https://stackoverflow.com/questions/24257803/distutilsoptionerror-must-supply-either-home-or-prefix-exec-prefix-not-both)~~ This should be fixed by not using `--target`, and as such the `lib` folder, on Macs. They don't require root privilege with Homebrew anyway

Twentysix26 committed:
[WIP] Red's launcher

- Interactive but can also be used with CLI flags
- startRed.bat / startRedLoop.bat functionality on all platforms
- Deals with requirements, installs them locally without the need of root
- Update options
- Downloads ffmpeg on Windows and verifies its files' integrity
- Maintenance options: can reset Red, its requiremenents and factory reset
- Generates batch files for fast start on Windows

Twentysix26 committed:
Change launch's update message

Twentysix26 committed:
Main menu reordering

Twentysix26 committed:
Handle KeyboardInterrupt while Red is running

ccubed commented:
Platform isn't a reliable check for  a 64bit run time.  All that tells you is that they're on a 64 bit platform not that they're running in the 64 bit run time space. See the note here https://docs.python.org/3.6/library/platform.html

You may not necessarily care about this, but it's a distinction.

palmtree5 commented:
Suggest changing
```python
if not PYTHON_OK:
    input("Red needs Python 3.5 or superior. Install the required "
          "version.\nPress enter to continue.")
```
to
```python
if not PYTHON_OK:
    print("Red needs Python 3.5 or superior. Install the required "
          "version.")
    exit(1)
```
Seems pointless to make the user input when we're just exiting anyway

palmtree5 commented:
So, issue:

The launcher hangs after exiting the bot process sometimes

The only situation I've found that causes this so far is playing audio, then stopping audio, then shutting down via the shutdown command

palmtree5 commented:
Also, regarding the WIP note, I presume that the `REQUIREMENTS` field applies to the cog-specific info.json file?

palmtree5 commented:
looks like all the resetting stuff works properly as well and I haven't run into any issues with the menuing. Didn't test cli flags stuff yet

Twentysix26 commented:
@palmtree5 Yes, cog specific

Twentysix26 committed:
Added `Bot.pip_install`

A thread safe way to install pip packages in the local lib folder. To be used mainly by the downloader cog.

Twentysix26 committed:
Wording

Twentysix26 committed:
[Downloader] Install requirements on cog install / update

I'm not fully satisfied with this implementation. See comments.

tekulvw commented:
Why not make use of `requirements.txt` at the root level of a cog repo instead of a key in `info.json`?

Kowlin commented:
If I can make a guess its because not all cogs require the full package of requirements.txt.

palmtree5 commented:
I'm inclined to agree that we should only install a req if the cog that needs it is installed

calebj commented on red.py:
```diff
@@ -174,6 +176,32 @@ def user_allowed(self, message):
         else:
             return True
 
+    async def pip_install(self, name, *, timeout=60):
```
Having a timeout here is unwise, something like PyNaCl will take at least that long to build.

Twentysix26 replied:
Do you suggest removing the timeout entirely?

Chovin replied:
could prompt the user when it takes too long. ask them to keep trying or kill it

palmtree5 commented:
Can we implement update notifications into the launcher somehow? Though that would likely only run when the launcher is started so there would need to be some plan for notifications after that

Twentysix26 committed:
Unloaded cogs should stay unloaded

tekulvw commented:
What about adding in stuff for sharding? Nothing needs to be implemented for it, just have a configurable # of shards in the menus that'll be implemented later

Twentysix26 committed:
Let user pick ffmpeg's bitness

Every single method I found to verify Windows' bitness can fail under specific circumstances. Better safe than sorry.

Twentysix26 commented:
Should be done. Needs some testing, especially the `[p]cog update` part

@tekulvw I don't think adding options for something that is not implemented yet is a good idea.
@palmtree5 I thought about implementing that, either before merging or in the future. Makes sense for it to be in the launcher.

Twentysix26 committed:
Refactoring, added 'data' reset option

Twentysix26 committed:
Add .git check

Chovin commented:
so um.
is there a list of what needs to be tested and what is already tested?
how do I test this given a short timeframe?

Chovin commented:
Lovely PR ❤️ 

Tested on Mac and here are some 
### points of discussion from discord:
1. breaks on any `pip` operation
![](https://cdn.discordapp.com/attachments/160386989819035648/264551270814056448/unknown.png)  
2. recommend letting the user know it's not done by saying something like "Updating requirements.." after it says "Red has been updated"
![](https://cdn.discordapp.com/attachments/160386989819035648/264552083934543873/unknown.png)
3. recommend saying "local installed python packages" or just take that part out, so that people don't think it uninstalls cogs.  
Maybe even point them to the `[p]cog` command as the source for cog management
![](https://cdn.discordapp.com/attachments/160386989819035648/264552734127030274/unknown.png)
4. > irdumb: nice catching smooth exit
irdumb: although, at times I'd like to restart it from discord without needing to have access to the console
26: maybe we should have a custom exit code for reboots. So that you can restart in any case
26: For technical reason, I propose 26

Twentysix26 committed:
Added `[p]restart`

Makes Red quit with exit code 26. This makes the launcher able to catch the request and restart Red on demand.
The method that this command calls has been integrated into the `Bot`class so that cogs can easily call it as well, if needed.
Both [p]shutdown and [p]restart now both respond in chat when issues but can be optionally silenced by passing a True bool (which, in discord.py, equals to "true", "yes", etc...)

Twentysix26 commented:
@Chovin  2, 3 and 4 are dealt with. Point 1 is the main issue now

Jan200101 commented:
What will happen when a requirement is already installed but not in the lib folder ?

Twentysix26 committed:
Refactoring

Twentysix26 commented:
@Jan200101 The `lib` folder has just priority over other paths. Nothing changes for existing installations. When they'll update with the launcher, requirements will get installed in the `lib` folder, taking priority over the default ones.

Jan200101 commented:
so we are basicly forced to redownload it into the lib folder even though we got it installed for the python version

Twentysix26 commented:
You are not "forced" to do anything. It's a minor inconvenience that happens the first time you decide to update using the *optional* launcher on existing installations. And a very minor inconvenience at that.

Chovin commented:
1. still seems to be an issue 
![image](https://cloud.githubusercontent.com/assets/17536161/21600412/72fc8962-d1c7-11e6-9e64-6749f06415b2.png)  


~~4. still seems to be an issue
![image](https://cloud.githubusercontent.com/assets/17536161/21600402/5273fa36-d1c7-11e6-80c8-96b8be5b7e4f.png)~~
[edit] didn't see the commit comment.. cause I don't look there :3 

`[p]restart` works
4 is all good 👍

Twentysix26 committed:
Don't use --target for Macs

Broken with Homebrew and doesn't seem necessary since it should default to the user's folder, not requiring root privilege

Twentysix26 committed:
No default timeout for pip installs

Chovin commented:
1. seems to be fixed 👍

Twentysix26 committed:
Generate fast start scripts for Unix too

Upon the launcher being started fast start scripts will be created (batch files / sh scripts)

Twentysix26 committed:
.command extension for mac scripts

Twentysix26 committed:
Spelling

Twentysix26 committed:
Add option to update pip

Twentysix26 committed:
Handle missing git error

Twentysix26 committed:
Inaccuracy
